node/bindnode/infer.go

```diff
@@ -206,7 +206,30 @@ func verifyCompatibility(seen map[seenEntry]bool, goType reflect.Type, schemaTyp
 	}
 }
 
-func inferGoType(typ schema.Type) reflect.Type {
+// If we recurse past a large number of levels, we're mostly stuck in a loop.
+// Prevent burning CPU or causing OOM crashes.
+// If a user really wrote an IPLD schema or Go type with such deep nesting,
+// it's likely they are trying to abuse the system as well.
+const maxRecursionLevel = 1 << 10
```

Comment:
1024 is still pretty large; find if this isn't a significant cost but you could also lower this and wait till someone squeals and asks for it to be increased

Reply:
Maybe - I admit I chose it arbitrarily. It's high enough that I don't expect anyone will actually run into it, but low enough that it should never cause an OOM on its own.